src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/groovy/SecureGroovyScript.java

```diff
@@ -151,36 +151,49 @@ public SecureGroovyScript configuringWithNonKeyItem() {
     }
 
     private static void cleanUpLoader(ClassLoader loader, Set<ClassLoader> encounteredLoaders, Set<Class<?>> encounteredClasses) throws Exception {
-        /*if (loader instanceof CpsGroovyShell.TimingLoader) {
-            cleanUpLoader(loader.getParent(), encounteredLoaders, encounteredClasses);
-            return;
-        }*/
-        // Check me, am I cleaning up the right loader???
-        if (!(loader instanceof GroovyClassLoader)) {
-            LOGGER.log(Level.FINER, "ignoring {0}", loader);
-            return;
-        }
         if (!encounteredLoaders.add(loader)) {
             return;
         }
-        cleanUpLoader(loader.getParent(), encounteredLoaders, encounteredClasses);
         if (LOGGER.isLoggable(Level.FINER)) {
           LOGGER.log(Level.FINER, "found {0}", String.valueOf(loader));
         }
-        cleanUpGlobalClassValue(loader);
-        GroovyClassLoader gcl = (GroovyClassLoader) loader;
-        for (Class<?> clazz : gcl.getLoadedClasses()) {
-            if (encounteredClasses.add(clazz)) {
-                LOGGER.log(Level.FINER, "found {0}", clazz.getName());
-                Introspector.flushFromCaches(clazz);
-                cleanUpGlobalClassSet(clazz);
-                cleanUpObjectStreamClassCaches(clazz);
-                cleanUpLoader(clazz.getClassLoader(), encounteredLoaders, encounteredClasses);
+        if (loader instanceof GroovyClassLoader) {
+            GroovyClassLoader gcl = (GroovyClassLoader) loader;
+            for (Class<?> clazz : gcl.getLoadedClasses()) {
+                cleanUpClass(clazz, encounteredLoaders, encounteredClasses);
+            }
+            gcl.clearCache();
+        } else if (loader instanceof SandboxResolvingClassLoader) {
+            // OK, just check its parent
+        } else if (loader instanceof ClasspathURLClassLoader) {
+            Collection<Class<?>> loadedClasses = ((ClasspathURLClassLoader) loader).loadedClasses;
+            synchronized (loadedClasses) {
+                loadedClasses = new ArrayList<>(loadedClasses);
```

Comment:
Do you need to call `ClasspathURLClassLoader.loadedClasses.clear()` here? It's not obvious to me why the collection of strong references to the classes in `ClasspathURLClassLoader` is not problematic. Is it because we are only concerned about external references in caches, etc., and so any references from the class loaders to the classes they loaded are fine so long as the class loader is not strongly referenced anywhere?

Reply:
`ClassLoader` already has

```java
// The classes loaded by this class loader. The only purpose of this table
// is to keep the classes from being GC'ed until the loader is GC'ed.
private final Vector<Class<?>> classes = new Vector<>();
```

but it is inaccessible, so we need a custom subclass that makes a similar list accessible. `GroovyClassLoader` happens to have such a method already, which we were using. Also having a custom subclass helps us identify a loader created earlier by `SecureGroovyScript` (with a nonempty classpath), so that we know when to terminate the recursive `parent` search (normally it ends at `UberClassLoader` but the API does permit a plugin to pass in some other base loader, typically a `PluginClassLoader`).